src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php

```diff
@@ -65,161 +67,74 @@ public function setVoters(array $voters)
     }
 
     /**
-     * {@inheritdoc}
+     * @param mixed $strategies
      */
-    public function decide(TokenInterface $token, array $attributes, $object = null)
+    public function addStrategy($name,$strategy)
     {
-        return $this->{$this->strategy}($token, $attributes, $object);
+        $this->strategies[$name] = $strategy;
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function supportsAttribute($attribute)
+    private function getStrategy($strategyName)
     {
-        foreach ($this->voters as $voter) {
-            if ($voter->supportsAttribute($attribute)) {
-                return true;
+        if(!array_key_exists($strategyName,$this->strategies))
+        {
+            switch($strategyName){
+                case self::STRATEGY_UNANIMOUS:
+                    return new DecideUnanimousStrategy();
+                case self::STRATEGY_CONSENSUS:
+                    return new DecideConsensusStrategy();
+                case self::STRATEGY_AFFIRMATIVE:
+                    return new DecideAffirmativeStrategy();
+                case self::STRATEGY_HIGHEST_NOT_ABSTAINED:
+                    return new DecideHighestNotAbstainedVoterStrategy();
+                default:
+                    break;
             }
+        } elseif($this->strategies[$strategyName] instanceof AccessDecisionStrategyInterface) {
+            return $this->strategies[$strategyName];
         }
 
-        return false;
+        throw new \InvalidArgumentException(sprintf('The strategy "%s" is not supported.', $strategyName));
     }
 
     /**
      * {@inheritdoc}
      */
-    public function supportsClass($class)
+    public function decide(TokenInterface $token, array $attributes, $object = null)
     {
-        foreach ($this->voters as $voter) {
-            if ($voter->supportsClass($class)) {
-                return true;
-            }
-        }
+        $strategy = $this->getStrategy($this->strategy);
+        $strategy->setVoters($this->voters);
+        $strategy->setAllowIfAllAbstainDecisions($this->allowIfAllAbstainDecisions);
+        $strategy->setAllowIfEqualGrantedDeniedDecisions($this->allowIfEqualGrantedDeniedDecisions);
```

Comment:
I'm not sure this is the desired way, why set it every decision? Why not just do it when building the container?